base/strings/string.jl

```diff
@@ -323,11 +323,11 @@ end
 codelen(c::Char) = 4 - (trailing_zeros(0xff000000 | reinterpret(UInt32, c)) >> 3)
 
 function string(a::Union{String,AbstractChar}...)
-    sprint() do io
-        for x in a
-            print(io, x)
-        end
+    io = IOBuffer()
+    for x in a
+        print(io, x)
     end
+    return String(resize!(io.data, io.size))
```

Comment:
Why not `String(take!(io))`?

Reply:
This consumes the data rather than making a copy.

Reply:
In this case, `take!` behaves the same way and doesn't make a copy.  Ah, but I do understand now, the issue is that `take!` also is preparing the `IOBuffer` for additional writes, and allocates a new buffer to prepare it for additional writes.

In Kristoffer's `@btime string("foo", 'b', "ar")`, this incurs a 30% penalty.  Alright, carry on.

Reply:
I wish `take!` didn't do that. In the vast majority (all?) of cases, an IOBuffer is not used after `take!` is called, so we should at least optimize for that case.

Reply:
Opened https://github.com/JuliaLang/julia/issues/27741